Removed ENI status property to fix logic issue

sas-pemcne opened the pull request:
I have removed the Status property from the EC2NetworkInterface resource to avoid issues where the nuker thinks the resource has been deleted when it has not. 

According to the logic of the nuker: 
1. Resources are triggered for deletion
2. All resources are queried again
3. If the resource in the query matches a resource in the initial cache, the resource still exists, otherwise it has been successfully deleted

A problem arises in the matching logic because all properties of the resource must be identical for the match to be successful. This dictates that all properties must be immutable for the nuker to function properly. The status property of a EC2NetworkInterface would change from "in-use" to "available" (assuming it was attached) during the deletion process. Because of that, the resource could not be matched (step 3 above) and the nuker prematurely assumes the ENI has been deleted when the delete command had not been successfully issued. 

This might help some issues related to VPC dependency errors like #503 since any attached ENI will not be removed correctly. I also realize this will cause some breaking changes if others have this filtered in the config but I couldn't think of an easier way to resolve the issue without overhauling the matching logic.

sas-pemcne committed:
Add String() and handle expected errors from the detach

ekristen commented:
Woah, great find, I was just reviewing the code and agree with your assessment, I actually think this behavior is bad. Properties shouldn't have to match especially when resources have unique ids and those properties can change. 

This might be a bug in the core queue handling that needs to be addressed.

svenwltr commented:
I agree with @ekristen.

The original idea was to deprecate the `String()` function for resources and use `Properties()` only, but this was never pushed forward. Therefore most of the resources still have this. I think making the `String()` function not optional anymore and using this as the primary key would solve all bugs of this kind. Also it might make sense to rename it to `ID()` or similar in another step, but that's not necessary.

For this particular resource I suggest adding the String() function and return the `NetworkInterfaceId` there.

sas-pemcne commented:
Good catch with the String() comparison, I was so focused on the Properties that I completely missed that check in the Equals function. I added the String() and local testing looks good. 

A byproduct of actually having the ENI being deleted correctly is it was complaining about the attachment not existing when it tries to remove it and the Remove() would throw that error back. I'm guessing this is because the resource itself is not updated as the nuker progresses so it still thinks it has an attachment when it in fact does not. That or I was hitting a very coincidental timing issue during the detaching process. I removed the error return just for the detachment process which matches the same logic used in the EC2SecurityGroups where the pre-removal steps are performed and the output is not logged. This allows the ENI to be properly removed since it can actually make it down to the delete call.

svenwltr commented on resources/ec2-network-interfaces.go:
```diff
@@ -37,14 +37,12 @@ func ListEC2NetworkInterfaces(sess *session.Session) ([]Resource, error) {
+
 	if e.eni.Attachment != nil {
-		_, err := e.svc.DetachNetworkInterface(&ec2.DetachNetworkInterfaceInput{
+		e.svc.DetachNetworkInterface(&ec2.DetachNetworkInterfaceInput{
```
Ignoring the error completely has the problem that it would assume the detachment worked properly on all other issues too (for example a connection issue).

Can we specifically check that the error indicates that the network interface is already gone?

sas-pemcne replied:
I added a check for the specific error and ignore it, otherwise if it is unexpected it will get returned back.
